src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -166,6 +179,221 @@ instance (NFData t, NFData r) => NFData (SnapMergingRunState t r) where
+  deriving newtype NFData
+
+data SnapMergingTreeState r =
+    SnapCompletedTreeMerge !r
+  | SnapPendingTreeMerge !(SnapPendingMerge r)
+  | SnapOngoingTreeMerge
+      !NumRuns
+      !NumEntries
+      !SuppliedCredits
+      !(SnapMergingRunState MR.TreeMergeType r)
+  deriving stock (Show, Eq, Functor, Foldable, Traversable)
+
+instance NFData r => NFData (SnapMergingTreeState r) where
+  rnf (SnapCompletedTreeMerge a) = rnf a
+  rnf (SnapPendingTreeMerge a) = rnf a
+  rnf (SnapOngoingTreeMerge a b c d) =
+      rnf a `seq` rnf b `seq` rnf c `seq` rnf d
+
+data SnapPendingMerge r =
+    SnapPendingLevelMerge
+      ![SnapPreExistingRun r]
+      !(Maybe (SnapMergingTree r))
+  | SnapPendingUnionMerge
+      ![SnapMergingTree r]
+  deriving stock (Show, Eq, Functor, Foldable, Traversable)
+
+instance NFData r => NFData (SnapPendingMerge r) where
+  rnf (SnapPendingLevelMerge a b) = rnf a `seq` rnf b
+  rnf (SnapPendingUnionMerge a)   = rnf a
+
+data SnapPreExistingRun r =
+    SnapPreExistingRun !r
+  | SnapPreExistingMergingRun
+      !NumRuns
+      !NumEntries
+      !SuppliedCredits
+      !(SnapMergingRunState MR.LevelMergeType r)
+  deriving stock (Show, Eq, Functor, Foldable, Traversable)
+
+instance NFData r => NFData (SnapPreExistingRun r) where
+  rnf (SnapPreExistingRun a) = rnf a
+  rnf (SnapPreExistingMergingRun a b c d) =
+      rnf a `seq` rnf b `seq` rnf c `seq` rnf d
+
+{-------------------------------------------------------------------------------
+  Opening from merging tree snapshot format
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE fromSnapMergingTree ::
+     ActionRegistry IO
+  -> HasFS IO h
+  -> HasBlockIO IO h
+  -> TableConfig
+  -> UniqCounter IO
+  -> ResolveSerialisedValue
+  -> ActiveDir
+  -> SnapMergingTree (Ref (Run IO h))
+  -> IO (Ref (MT.MergingTree IO h))
+  #-}
+-- | Duplicates runs and re-creates merging runs.
+fromSnapMergingTree ::
+     forall m h. (MonadMask m, MonadMVar m, MonadSTM m, MonadST m)
+  => ActionRegistry m
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> TableConfig
+  -> UniqCounter m
+  -> ResolveSerialisedValue
+  -> ActiveDir
+  -> SnapMergingTree (Ref (Run m h))
+  -> m (Ref (MT.MergingTree m h))
+fromSnapMergingTree reg hfs hbio conf@TableConfig{..} uc resolve dir (SnapMergingTree snapTreeState) =
+    fromSnapTreeState snapTreeState
+  where
+    mkPath = RunFsPaths (getActiveDir dir)
+
+    dupRun :: RefCounted m obj => Ref obj -> m (Ref obj)
+    dupRun r = withRollback reg (dupRef r) releaseRef
+
+    -- Used for inductive calls
+    recurrance :: SnapMergingTree (Ref (Run m h)) -> m (Ref (MT.MergingTree m h))
+    recurrance = fromSnapMergingTree reg hfs hbio conf uc resolve dir
+
+    fromSnapTreeState :: SnapMergingTreeState (Ref (Run m h)) -> m (Ref (MT.MergingTree m h))
+    fromSnapTreeState (SnapCompletedTreeMerge run) = dupRun run >>= \run' ->
+      MT.newPendingLevelMerge [MT.PreExistingRun run'] Nothing
```

Comment:
Hmm, is duping the right thing here? The `newPendingLevelMerge` will make its own reference duplicates, we should not need to duplicate them first.

Where would these duplicate references be owned or released?

I'm a bit surprised that this passes tests without throwing exceptions for forgotten references.